llvm/lib/Transforms/Utils/SCCPSolver.cpp

```diff
@@ -82,6 +83,26 @@ bool SCCPSolver::tryToReplaceWithConstant(Value *V) {
     return false;
   }
 
+  // Perform constant pointer propagation as long as assuming PredicateInfo
+  // derived equality between the two holds, and their provenance is the same.
+  if (auto *I = dyn_cast<Instruction>(V); I && I->getType()->isPointerTy()) {
+    if (getPredicateInfoFor(I)) {
```

Comment:
Won't think not work if there any intermediate copies of the value? Like through a phi/select or in the IPSCCP case a function return value etc.

Reply:
On second thought, I believe there's not much we can do but visit the transitive closure of an instruction's operands, see whether the copy is part of it.